src/tdigest_api.c

```diff
@@ -0,0 +1,693 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+/*
+ * T-digest-based aggregate and distribution functions
+ *
+ * Originally from PipelineDB 1.0.0
+ * Original copyright (c) 2018, PipelineDB, Inc. and released under Apache License 2.0
+ * Modifications copyright by Timescale, Inc. per NOTICE
+ */
+
+#include "postgres.h"
+
+#include "tdigest.h"
+#include "tdigest_api.h"
+#include "fmgr.h"
+#include "export.h"
+
+#include "lib/stringinfo.h"
+#include "libpq/pqformat.h"
+#include "nodes/nodeFuncs.h"
+#include "utils/builtins.h"
+#include "utils/bytea.h"
+#include "utils/datum.h"
+#include "utils/typcache.h"
+#include "compat.h"
+
+/*
+ * Creates string representation of TDigest
+ */
+static char *
+ts_tdigest_print(TDigest *t)
+{
+	StringInfoData buf;
+
+	initStringInfo(&buf);
+	appendStringInfo(&buf,
+					 "{ count = %ld, k = %d, centroids: %d, size = %dB }",
+					 t->total_weight,
+					 (int) t->compression,
+					 t->num_centroids,
+					 (int) ts_tdigest_size(t));
+
+	return buf.data;
+}
+
+/*
+ * Returns string representation of TDigest in a transition struct
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_print_sql);
+Datum
+ts_tdigest_print_sql(PG_FUNCTION_ARGS)
+{
+	TDigest *t;
+	TDigestTState *tts;
+
+	if (PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+
+	tts = (TDigestTState *) PG_GETARG_VARLENA_P(0);
+	t = tts->td;
+
+	PG_RETURN_TEXT_P(CStringGetTextDatum(ts_tdigest_print(t)));
+}
+
+/*
+ * Sets up a TDigest aggregate
+ */
+static TDigestTState *
+ts_tdigest_startup(uint32_t k)
+{
+	TDigestTState *tts;
+	TDigest *t;
+
+	tts = palloc0(sizeof(TDigestTState));
+
+	// create tdigest
+	if (k)
+		t = ts_tdigest_create_with_compression(k);
+	else
+		t = ts_tdigest_create();
+
+	// configure fields
+	tts->td = t;
+	tts->unmerged_centroids = NIL;
+
+	return tts;
+}
+
+/*
+ * Compresses buffered data in a TDigest aggregate into the TDigest
+ * This is the only API function that should directly call TDigest's compress function
```

Comment:
Doing so would require bringing one of the `tdigest` methods into the API file; rationale in keeping them separate is to keep the SQL-facing functions and the actual T-Digest logic decoupled.

Reply:
I thought to move this function into T-Digist file, but it is not a good idea, since it will bring T-Digist-State there. So good point and it is better to keep as now.